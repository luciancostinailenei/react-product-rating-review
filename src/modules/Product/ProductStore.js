import { action, computed, observable } from 'mobx'

import moment from 'moment'

class ProductStore {
  @observable reviews = []
  @observable isProductReviewFormDisplayed = false

  @action
  addReview = review => {
    this.reviews.push(review)
  }

  @computed
  get reviewsCount() {
    return this.reviews.length
  }

  @computed
  get sortedReviews() {
    const sortedReviews = this.reviews.sort((a, b) => {
      if (moment(a.date) > moment(b.date)) {
        return -1
      }

      if (moment(a.date) < moment(b.date)) {
        return 1
      }

      return 0
    })

    return sortedReviews
  }

  @computed
  get averageScore() {
    const numberOfReviews = this.reviewsCount

    if (numberOfReviews === 0) {
      return 0
    }

    const scoreSum = this.reviews.reduce((acc, review) => {
      return acc + review.rating
    }, 0)

    const averageScore = scoreSum / numberOfReviews
    return averageScore % 1 === 0 ? averageScore : averageScore.toFixed(1)
  }
}

const store = new ProductStore()
export default store
