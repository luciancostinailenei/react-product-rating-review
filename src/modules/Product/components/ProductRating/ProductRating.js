import React, { Component } from 'react'
import { inject, observer } from 'mobx-react'
import PropTypes from 'prop-types'

import './ProductRating.scss'

@inject('ProductStore')
@observer
export default class ProductRating extends Component {
  constructor(props) {
    super(props)

    this.state = {
      checkedStarId: null
    }

    this.resetVote = this.resetVote.bind(this)
  }

  changeRating(id) {
    if (this.props.scoreValue) {
      return
    }

    this.setState({ checkedStarId: id })

    this.props.changeRatingHandler(parseInt(id, 10))
  }

  isStarChecked(id) {
    if (this.props.scoreValue) {
      return this.props.scoreValue === parseInt(id, 10)
    }

    return this.state.checkedStarId === id
  }

  resetVote() {
    this.setState({ checkedStarId: null })
  }

  render() {
    return (
      <span
        className={
          'ProductRating' +
          (this.props.scoreValue ? ' ProductRating--disabled' : '')
        }
      >
        <input
          onChange={() => {
            this.changeRating('5')
          }}
          value={this.props.inputIdIndex}
          type="radio"
          className="ProductRating__input"
          id={'ProductRating__input-' + this.props.inputIdIndex + '-5'}
          name={'ProductRating__input-' + this.props.inputIdIndex}
          checked={this.isStarChecked('5')}
        />
        <label
          htmlFor={'ProductRating__input-' + this.props.inputIdIndex + '-5'}
          className="ProductRating__star"
        />

        <input
          onChange={() => {
            this.changeRating('4')
          }}
          value={this.props.inputIdIndex}
          type="radio"
          className="ProductRating__input"
          id={'ProductRating__input-' + this.props.inputIdIndex + '-4'}
          name={'ProductRating__input-' + this.props.inputIdIndex}
          checked={this.isStarChecked('4') ? true : false}
        />
        <label
          htmlFor={'ProductRating__input-' + this.props.inputIdIndex + '-4'}
          className="ProductRating__star"
        />

        <input
          onChange={() => {
            this.changeRating('3')
          }}
          value={this.props.inputIdIndex}
          type="radio"
          className="ProductRating__input"
          id={'ProductRating__input-' + this.props.inputIdIndex + '-3'}
          name={'ProductRating__input-' + this.props.inputIdIndex}
          checked={this.isStarChecked('3')}
        />
        <label
          htmlFor={'ProductRating__input-' + this.props.inputIdIndex + '-3'}
          className="ProductRating__star"
        />

        <input
          onChange={() => {
            this.changeRating('2')
          }}
          value={this.props.inputIdIndex}
          type="radio"
          className="ProductRating__input"
          id={'ProductRating__input-' + this.props.inputIdIndex + '-2'}
          name={'ProductRating__input-' + this.props.inputIdIndex}
          checked={this.isStarChecked('2')}
        />
        <label
          htmlFor={'ProductRating__input-' + this.props.inputIdIndex + '-2'}
          className="ProductRating__star"
        />

        <input
          onChange={() => {
            this.changeRating('1')
          }}
          value={this.props.inputIdIndex}
          type="radio"
          className="ProductRating__input"
          id={'ProductRating__input-' + this.props.inputIdIndex + '-1'}
          name={'ProductRating__input-' + this.props.inputIdIndex}
          checked={this.isStarChecked('1')}
        />
        <label
          htmlFor={'ProductRating__input-' + this.props.inputIdIndex + '-1'}
          className="ProductRating__star"
        />
      </span>
    )
  }
}

ProductRating.propTypes = {
  scoreValue: PropTypes.number,
  inputIdIndex: PropTypes.string
}
