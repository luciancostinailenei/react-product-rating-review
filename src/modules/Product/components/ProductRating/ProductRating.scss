.ProductRating {
    overflow: hidden;
    display: inline-block;

    .ProductReviewForm & {
        margin-left: 5px;
    }
}

    .ProductRating__input {
        float: right;
        width: 16px;
        height: 16px;
        padding: 0;
        margin: 0 0 0 -16px;
        opacity: 0;

        &:checked {
            &~ .ProductRating__star {
                background-position: 0 0; 
            }
        }
    }

    .ProductRating__star {
        cursor: pointer;
        float: right;
        position: relative;
        display: block;
        width: 16px;
        height: 16px;
        background: url('/assets/star.png') 0 -16px;

        &:hover {
            background-position: 0 0;

            &~ .ProductRating__star {
                background-position: 0 0;
            }

            .ProductRating:hover & {
                background-position: 0 0;

                &~ .ProductRating__star {
                    background-position: 0 0;
                }
            }
        }

        .ProductRating__input:checked & {
            background-position: 0 0;
        }

        .ProductRating--disabled & { 
            pointer-events: none;
        }
    }


.ProductRating--disabled { }