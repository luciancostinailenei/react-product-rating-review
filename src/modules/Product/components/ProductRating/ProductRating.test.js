import { configure, mount, shallow } from 'enzyme'
import React from 'react'
import Adapter from 'enzyme-adapter-react-16'
import toJson, { createSerializer } from 'enzyme-to-json'

import ProductRating from './ProductRating'

import ProductStore from '../../ProductStore'

expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }))

configure({ adapter: new Adapter() })

describe('ProductRating', () => {
  it('matches snapshot on both active and disabled states', () => {
    const ProductRatingActiveWrapper = shallow(
      <ProductRating ProductStore={ProductStore} inputIdIndex="100" />
    )

    expect(toJson(ProductRatingActiveWrapper)).toMatchSnapshot()

    const ProductRatingDisabledWrapper = shallow(
      <ProductRating
        ProductStore={ProductStore}
        inputIdIndex="1"
        scoreValue={4}
      />
    )

    expect(toJson(ProductRatingDisabledWrapper)).toMatchSnapshot()
  })

  it('checks stars on user interaction', () => {
    const ProductRatingActiveWrapper = shallow(
      <ProductRating
        ProductStore={ProductStore}
        inputIdIndex="100"
        changeRatingHandler={() => 1}
      />
    ).dive()

    let firstStar = ProductRatingActiveWrapper.find(
      '#ProductRating__input-100-1'
    )

    firstStar.prop('onChange')()

    ProductRatingActiveWrapper.update()

    firstStar = ProductRatingActiveWrapper.find('#ProductRating__input-100-1')

    expect(firstStar.prop('checked')).toEqual(true)
  })

  it('displays checked stars based on given score prop value', () => {
    const ProductRatingDisabledWrapper = shallow(
      <ProductRating
        ProductStore={ProductStore}
        inputIdIndex="1"
        scoreValue={1}
      />
    )

    ProductRatingDisabledWrapper.update()

    expect(
      ProductRatingDisabledWrapper.dive()
        .find('#ProductRating__input-1-1')
        .prop('checked')
    ).toEqual(true)
  })
})
