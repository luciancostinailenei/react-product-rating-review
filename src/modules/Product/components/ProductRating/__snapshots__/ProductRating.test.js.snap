// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductRating matches snapshot on both active and disabled states 1`] = `
<ProductRating
  ProductStore={ProductStore {}}
  inputIdIndex="100"
/>
`;

exports[`ProductRating matches snapshot on both active and disabled states 2`] = `
<ProductRating
  ProductStore={ProductStore {}}
  inputIdIndex="1"
  scoreValue={4}
/>
`;
